ddtrace/profiling/_nogevent.py

```diff
@@ -0,0 +1,65 @@
+# -*- encoding: utf-8 -*-
+"""This files exposes non-gevent Python original functions."""
+import threading
+
+from ddtrace.vendor import six
+from ddtrace.vendor import attr
+
+
+try:
+    import gevent.monkey
+except ImportError:
+
+    def get_original(module, func):
+        return getattr(__import__(module), func)
+
+    def is_module_patched(module):
+        return False
+
+
+else:
+    get_original = gevent.monkey.get_original
+    is_module_patched = gevent.monkey.is_module_patched
+
+
+sleep = get_original("time", "sleep")
+
+try:
+    # Python ≥ 3.8
+    threading_get_native_id = get_original("threading", "get_native_id")
+except AttributeError:
+    threading_get_native_id = None
+
+start_new_thread = get_original(six.moves._thread.__name__, "start_new_thread")
+thread_get_ident = get_original(six.moves._thread.__name__, "get_ident")
+Thread = get_original("threading", "Thread")
+Lock = get_original("threading", "Lock")
+
+
+if is_module_patched("threading"):
+
+    @attr.s
+    class DoubleLock(object):
```

Comment:
:+1: I was skeptical of the need to lock across both coroutine and thread at first but I think I was able to work through it in my head (with my limited knowledge of gevent).

Reply:
Yeah, it's really not that obvious. Basically, if you have a thread with coroutines, you have concurrency in that thread between coroutines so you need a gevent-lock to make sure they are excluded and scheduled correctly.
The coroutine also needs to exclude threads, so you need a thread-lock to make sure you have, in the end, one thread and one coroutine acting.

I think what I was scared of is that you can use a gevent-lock from 2 threads, but it seems to work fine if you use a thread-lock first to synchronize. Scary!

```python
import threading
import sys
import gevent.monkey
import time

gevent.monkey.patch_all()

gl = threading.Lock()

sleep = gevent.monkey.get_original("time", "sleep")
thread = gevent.monkey.get_original("_thread", "start_new_thread")
tl = gevent.monkey.get_original("threading", "Lock")()


def lock_unlock():
    while True:
        tl.acquire()
        gl.acquire()
        print("thread got it")
        sleep(0.01)
        tl.release()
        gl.release()
        sleep(0.01)


th = thread(lock_unlock, ())

while True:
    tl.acquire()
    gl.acquire()
    print("gevent got it")
    time.sleep(0.01)
    tl.release()
    gl.release()
    sleep(0.01)
```

If you remove the thread-lock, it explodes.

I should write a blog post about that. 🤔

Reply:
Yeah I was curious about the ordering! Thanks for adding the comment here and in the source below 😄. Blog post would be great!